src/router/krb5/src/man/kdb5_util.man
.\" Man page generated from reStructuredText.
.
.TH "KDB5_UTIL" "8" " " "1.17" "MIT Kerberos"
.SH NAME
kdb5_util \- Kerberos database maintenance utility
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.sp
\fBkdb5_util\fP
[\fB\-r\fP \fIrealm\fP]
[\fB\-d\fP \fIdbname\fP]
[\fB\-k\fP \fImkeytype\fP]
[\fB\-kv\fP \fImkeyVNO\fP]
[\fB\-M\fP \fImkeyname\fP]
[\fB\-m\fP]
[\fB\-sf\fP \fIstashfilename\fP]
[\fB\-P\fP \fIpassword\fP]
[\fB\-x\fP \fIdb_args\fP]
\fIcommand\fP [\fIcommand_options\fP]
.SH DESCRIPTION
.sp
kdb5_util allows an administrator to perform maintenance procedures on
the KDC database.  Databases can be created, destroyed, and dumped to
or loaded from ASCII files.  kdb5_util can create a Kerberos master
key stash file or perform live rollover of the master key.
.sp
When kdb5_util is run, it attempts to acquire the master key and open
the database.  However, execution continues regardless of whether or
not kdb5_util successfully opens the database, because the database
may not exist yet or the stash file may be corrupt.
.sp
Note that some KDC database modules may not support all kdb5_util
commands.
.SH COMMAND-LINE OPTIONS
.INDENT 0.0
.TP
\fB\-r\fP \fIrealm\fP
specifies the Kerberos realm of the database.
.TP
\fB\-d\fP \fIdbname\fP
specifies the name under which the principal database is stored;
by default the database is that listed in kdc.conf(5)\&.  The
password policy database and lock files are also derived from this
value.
.TP
\fB\-k\fP \fImkeytype\fP
specifies the key type of the master key in the database.  The
default is given by the \fBmaster_key_type\fP variable in
kdc.conf(5)\&.
.TP
\fB\-kv\fP \fImkeyVNO\fP
Specifies the version number of the master key in the database;
the default is 1.  Note that 0 is not allowed.
.TP
\fB\-M\fP \fImkeyname\fP
principal name for the master key in the database.  If not
specified, the name is determined by the \fBmaster_key_name\fP
variable in kdc.conf(5)\&.
.TP
\fB\-m\fP
specifies that the master database password should be read from
the keyboard rather than fetched from a file on disk.
.TP
\fB\-sf\fP \fIstash_file\fP
specifies the stash filename of the master database password.  If
not specified, the filename is determined by the
\fBkey_stash_file\fP variable in kdc.conf(5)\&.
.TP
\fB\-P\fP \fIpassword\fP
specifies the master database password.  Using this option may
expose the password to other users on the system via the process
list.
.TP
\fB\-x\fP \fIdb_args\fP
specifies database\-specific options.  See kadmin(1) for
supported options.
.UNINDENT
.SH COMMANDS
.SS create
.INDENT 0.0
.INDENT 3.5
\fBcreate\fP [\fB\-s\fP]
.UNINDENT
.UNINDENT
.sp
Creates a new database.  If the \fB\-s\fP option is specified, the stash
file is also created.  This command fails if the database already
exists.  If the command is successful, the database is opened just as
if it had already existed when the program was first run.
.SS destroy
.INDENT 0.0
.INDENT 3.5
\fBdestroy\fP [\fB\-f\fP]
.UNINDENT
.UNINDENT
.sp
Destroys the database, first overwriting the disk sectors and then
unlinking the files, after prompting the user for confirmation.  With
the \fB\-f\fP argument, does not prompt the user.
.SS stash
.INDENT 0.0
.INDENT 3.5
\fBstash\fP [\fB\-f\fP \fIkeyfile\fP]
.UNINDENT
.UNINDENT
.sp
Stores the master principal\(aqs keys in a stash file.  The \fB\-f\fP
argument can be used to override the \fIkeyfile\fP specified in
kdc.conf(5)\&.
.SS dump
.INDENT 0.0
.INDENT 3.5
\fBdump\fP [\fB\-b7\fP|\fB\-ov\fP|\fB\-r13\fP|\fB\-r18\fP]
[\fB\-verbose\fP] [\fB\-mkey_convert\fP] [\fB\-new_mkey_file\fP
\fImkey_file\fP] [\fB\-rev\fP] [\fB\-recurse\fP] [\fIfilename\fP
[\fIprincipals\fP\&...]]
.UNINDENT
.UNINDENT
.sp
Dumps the current Kerberos and KADM5 database into an ASCII file.  By
default, the database is dumped in current format, "kdb5_util
load_dump version 7".  If filename is not specified, or is the string
"\-", the dump is sent to standard output.  Options:
.INDENT 0.0
.TP
\fB\-b7\fP
causes the dump to be in the Kerberos 5 Beta 7 format ("kdb5_util
load_dump version 4").  This was the dump format produced on
releases prior to 1.2.2.
.TP
\fB\-ov\fP
causes the dump to be in "ovsec_adm_export" format.
.TP
\fB\-r13\fP
causes the dump to be in the Kerberos 5 1.3 format ("kdb5_util
load_dump version 5").  This was the dump format produced on
releases prior to 1.8.
.TP
\fB\-r18\fP
causes the dump to be in the Kerberos 5 1.8 format ("kdb5_util
load_dump version 6").  This was the dump format produced on
releases prior to 1.11.
.TP
\fB\-verbose\fP
causes the name of each principal and policy to be printed as it
is dumped.
.TP
\fB\-mkey_convert\fP
prompts for a new master key.  This new master key will be used to
re\-encrypt principal key data in the dumpfile.  The principal keys
themselves will not be changed.
.TP
\fB\-new_mkey_file\fP \fImkey_file\fP
the filename of a stash file.  The master key in this stash file
will be used to re\-encrypt the key data in the dumpfile.  The key
data in the database will not be changed.
.TP
\fB\-rev\fP
dumps in reverse order.  This may recover principals that do not
dump normally, in cases where database corruption has occurred.
.TP
\fB\-recurse\fP
causes the dump to walk the database recursively (btree only).
This may recover principals that do not dump normally, in cases
where database corruption has occurred.  In cases of such
corruption, this option will probably retrieve more principals
than the \fB\-rev\fP option will.
.sp
Changed in version 1.15: Release 1.15 restored the functionality of the \fB\-recurse\fP
option.

.sp
Changed in version 1.5: The \fB\-recurse\fP option ceased working until release 1.15,
doing a normal dump instead of a recursive traversal.

.UNINDENT
.SS load
.INDENT 0.0
.INDENT 3.5
\fBload\fP [\fB\-b7\fP|\fB\-ov\fP|\fB\-r13\fP|\fB\-r18\fP] [\fB\-hash\fP]
[\fB\-verbose\fP] [\fB\-update\fP] \fIfilename\fP
.UNINDENT
.UNINDENT
.sp
Loads a database dump from the named file into the named database.  If
no option is given to determine the format of the dump file, the
format is detected automatically and handled as appropriate.  Unless
the \fB\-update\fP option is given, \fBload\fP creates a new database
containing only the data in the dump file, overwriting the contents of
any previously existing database.  Note that when using the LDAP KDC
database module, the \fB\-update\fP flag is required.
.sp
Options:
.INDENT 0.0
.TP
\fB\-b7\fP
requires the database to be in the Kerberos 5 Beta 7 format
("kdb5_util load_dump version 4").  This was the dump format
produced on releases prior to 1.2.2.
.TP
\fB\-ov\fP
requires the database to be in "ovsec_adm_import" format.  Must be
used with the \fB\-update\fP option.
.TP
\fB\-r13\fP
requires the database to be in Kerberos 5 1.3 format ("kdb5_util
load_dump version 5").  This was the dump format produced on
releases prior to 1.8.
.TP
\fB\-r18\fP
requires the database to be in Kerberos 5 1.8 format ("kdb5_util
load_dump version 6").  This was the dump format produced on
releases prior to 1.11.
.TP
\fB\-hash\fP
stores the database in hash format, if using the DB2 database
type.  If this option is not specified, the database will be
stored in btree format.  This option is not recommended, as
databases stored in hash format are known to corrupt data and lose
principals.
.TP
\fB\-verbose\fP
causes the name of each principal and policy to be printed as it
is dumped.
.TP
\fB\-update\fP
records from the dump file are added to or updated in the existing
database.  Otherwise, a new database is created containing only
what is in the dump file and the old one destroyed upon successful
completion.
.UNINDENT
.SS ark
.INDENT 0.0
.INDENT 3.5
\fBark\fP [\fB\-e\fP \fIenc\fP:\fIsalt\fP,...] \fIprincipal\fP
.UNINDENT
.UNINDENT
.sp
Adds new random keys to \fIprincipal\fP at the next available key version
number.  Keys for the current highest key version number will be
preserved.  The \fB\-e\fP option specifies the list of encryption and
salt types to be used for the new keys.
.SS add_mkey
.INDENT 0.0
.INDENT 3.5
\fBadd_mkey\fP [\fB\-e\fP \fIetype\fP] [\fB\-s\fP]
.UNINDENT
.UNINDENT
.sp
Adds a new master key to the master key principal, but does not mark
it as active.  Existing master keys will remain.  The \fB\-e\fP option
specifies the encryption type of the new master key; see
Encryption_types in kdc.conf(5) for a list of possible
values.  The \fB\-s\fP option stashes the new master key in the stash
file, which will be created if it doesn\(aqt already exist.
.sp
After a new master key is added, it should be propagated to replica
servers via a manual or periodic invocation of kprop(8)\&.  Then,
the stash files on the replica servers should be updated with the
kdb5_util \fBstash\fP command.  Once those steps are complete, the key
is ready to be marked active with the kdb5_util \fBuse_mkey\fP command.
.SS use_mkey
.INDENT 0.0
.INDENT 3.5
\fBuse_mkey\fP \fImkeyVNO\fP [\fItime\fP]
.UNINDENT
.UNINDENT
.sp
Sets the activation time of the master key specified by \fImkeyVNO\fP\&.
Once a master key becomes active, it will be used to encrypt newly
created principal keys.  If no \fItime\fP argument is given, the current
time is used, causing the specified master key version to become
active immediately.  The format for \fItime\fP is getdate string.
.sp
After a new master key becomes active, the kdb5_util
\fBupdate_princ_encryption\fP command can be used to update all
principal keys to be encrypted in the new master key.
.SS list_mkeys
.INDENT 0.0
.INDENT 3.5
\fBlist_mkeys\fP
.UNINDENT
.UNINDENT
.sp
List all master keys, from most recent to earliest, in the master key
principal.  The output will show the kvno, enctype, and salt type for
each mkey, similar to the output of kadmin(1) \fBgetprinc\fP\&.  A
\fB*\fP following an mkey denotes the currently active master key.
.SS purge_mkeys
.INDENT 0.0
.INDENT 3.5
\fBpurge_mkeys\fP [\fB\-f\fP] [\fB\-n\fP] [\fB\-v\fP]
.UNINDENT
.UNINDENT
.sp
Delete master keys from the master key principal that are not used to
protect any principals.  This command can be used to remove old master
keys all principal keys are protected by a newer master key.
.INDENT 0.0
.TP
\fB\-f\fP
does not prompt for confirmation.
.TP
\fB\-n\fP
performs a dry run, showing master keys that would be purged, but
not actually purging any keys.
.TP
\fB\-v\fP
gives more verbose output.
.UNINDENT
.SS update_princ_encryption
.INDENT 0.0
.INDENT 3.5
\fBupdate_princ_encryption\fP [\fB\-f\fP] [\fB\-n\fP] [\fB\-v\fP]
[\fIprinc\-pattern\fP]
.UNINDENT
.UNINDENT
.sp
Update all principal records (or only those matching the
\fIprinc\-pattern\fP glob pattern) to re\-encrypt the key data using the
active database master key, if they are encrypted using a different
version, and give a count at the end of the number of principals
updated.  If the \fB\-f\fP option is not given, ask for confirmation
before starting to make changes.  The \fB\-v\fP option causes each
principal processed to be listed, with an indication as to whether it
needed updating or not.  The \fB\-n\fP option performs a dry run, only
showing the actions which would have been taken.
.SS tabdump
.INDENT 0.0
.INDENT 3.5
\fBtabdump\fP [\fB\-H\fP] [\fB\-c\fP] [\fB\-e\fP] [\fB\-n\fP] [\fB\-o\fP \fIoutfile\fP]
\fIdumptype\fP
.UNINDENT
.UNINDENT
.sp
Dump selected fields of the database in a tabular format suitable for
reporting (e.g., using traditional Unix text processing tools) or
importing into relational databases.  The data format is tab\-separated
(default), or optionally comma\-separated (CSV), with a fixed number of
columns.  The output begins with a header line containing field names,
unless suppression is requested using the \fB\-H\fP option.
.sp
The \fIdumptype\fP parameter specifies the name of an output table (see
below).
.sp
Options:
.INDENT 0.0
.TP
\fB\-H\fP
suppress writing the field names in a header line
.TP
\fB\-c\fP
use comma separated values (CSV) format, with minimal quoting,
instead of the default tab\-separated (unquoted, unescaped) format
.TP
\fB\-e\fP
write empty hexadecimal string fields as empty fields instead of
as "\-1".
.TP
\fB\-n\fP
produce numeric output for fields that normally have symbolic
output, such as enctypes and flag names.  Also requests output of
time stamps as decimal POSIX time_t values.
.TP
\fB\-o\fP \fIoutfile\fP
write the dump to the specified output file instead of to standard
output
.UNINDENT
.sp
Dump types:
.INDENT 0.0
.TP
\fBkeydata\fP
principal encryption key information, including actual key data
(which is still encrypted in the master key)
.INDENT 7.0
.TP
\fBname\fP
principal name
.TP
\fBkeyindex\fP

src/router/krb5/src/kadmin/dbutil/kdb5_util.c
/* -*- mode: c; c-basic-offset: 4; indent-tabs-mode: nil -*- */
/* kadmin/dbutil/kdb5_util.c - Administer a KDC database */
/*
 * (C) Copyright 1990,1991, 1996, 2008, 2009 by the Massachusetts Institute of Technology.
 * All Rights Reserved.
 *
 * Export of this software from the United States of America may
 *   require a specific license from the United States Government.
 *   It is the responsibility of any person or organization contemplating
 *   export to obtain such a license before exporting.
 *
 * WITHIN THAT CONSTRAINT, permission to use, copy, modify, and
 * distribute this software and its documentation for any purpose and
 * without fee is hereby granted, provided that the above copyright
 * notice appear in all copies and that both that copyright notice and
 * this permission notice appear in supporting documentation, and that
 * the name of M.I.T. not be used in advertising or publicity pertaining
 * to distribution of the software without specific, written prior
 * permission.  Furthermore if you modify this software you must label
 * your software as modified software and not distribute it in such a
 * fashion that it might be confused with the original M.I.T. software.
 * M.I.T. makes no representations about the suitability of
 * this software for any purpose.  It is provided "as is" without express
 * or implied warranty.
 */
/*
 * Copyright (C) 1998 by the FundsXpress, INC.
 *
 * All rights reserved.
 *
 * Export of this software from the United States of America may require
 * a specific license from the United States Government.  It is the
 * responsibility of any person or organization contemplating export to
 * obtain such a license before exporting.
 *
 * WITHIN THAT CONSTRAINT, permission to use, copy, modify, and
 * distribute this software and its documentation for any purpose and
 * without fee is hereby granted, provided that the above copyright
 * notice appear in all copies and that both that copyright notice and
 * this permission notice appear in supporting documentation, and that
 * the name of FundsXpress. not be used in advertising or publicity pertaining
 * to distribution of the software without specific, written prior
 * permission.  FundsXpress makes no representations about the suitability of
 * this software for any purpose.  It is provided "as is" without express
 * or implied warranty.
 *
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED
 * WARRANTIES OF MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.
 */
/*
 * Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 */

#include <k5-int.h>
#include <kadm5/admin.h>
#include <locale.h>
#include <adm_proto.h>
#include <time.h>
#include "kdb5_util.h"

/*
 * XXX Ick, ick, ick.  These global variables shouldn't be global....
 */
char *mkey_password = 0;

/*
 * I can't figure out any way for this not to be global, given how ss
 * works.
 */

int exit_status = 0;
krb5_context util_context;
kadm5_config_params global_params;

void usage()
{
    fprintf(stderr,
            _("Usage: kdb5_util [-r realm] [-d dbname] "
              "[-k mkeytype] [-kv mkeyVNO]\n"
              "\t        [-M mkeyname] [-m] [-sf stashfilename] "
              "[-P password]\n"
              "\t        [-x db_args]* cmd [cmd_options]\n"
              "\tcreate  [-s]\n"
              "\tdestroy [-f]\n"
              "\tstash   [-f keyfile]\n"
              "\tdump    [-old|-ov|-b6|-b7|-r13|-r18] [-verbose]\n"
              "\t        [-mkey_convert] [-new_mkey_file mkey_file]\n"
              "\t        [-rev] [-recurse] [filename [princs...]]\n"
              "\tload    [-old|-ov|-b6|-b7|-r13|-r18] [-verbose] [-update] "
              "filename\n"
              "\tark     [-e etype_list] principal\n"
              "\tadd_mkey [-e etype] [-s]\n"
              "\tuse_mkey kvno [time]\n"
              "\tlist_mkeys\n"));
    /* avoid a string length compiler warning */
    fprintf(stderr,
            _("\tupdate_princ_encryption [-f] [-n] [-v] [princ-pattern]\n"
              "\tpurge_mkeys [-f] [-n] [-v]\n"
              "\ttabdump [-H] [-c] [-e] [-n] [-o outfile] dumptype\n"
              "\nwhere,\n\t[-x db_args]* - any number of database specific "
              "arguments.\n"
              "\t\t\tLook at each database documentation for supported "
              "arguments\n"));
    exit(1);
}

krb5_keyblock master_keyblock;
krb5_kvno   master_kvno; /* fetched */
extern krb5_principal master_princ;
char *mkey_fullname;
krb5_db_entry *master_entry = NULL;
int     valid_master_key = 0;

char *progname;
krb5_boolean manual_mkey = FALSE;
krb5_boolean dbactive = FALSE;

static int open_db_and_mkey(void);

static void add_random_key(int, char **);

typedef void (*cmd_func)(int, char **);

struct _cmd_table {
    char *name;
    cmd_func func;
    int opendb;
} cmd_table[] = {
    {"create", kdb5_create, 0},
    {"destroy", kdb5_destroy, 1}, /* 1 opens the kdb */
    {"stash", kdb5_stash, 1},
    {"dump", dump_db, 1},
    {"load", load_db, 0},
    {"ark", add_random_key, 1},
    {"add_mkey", kdb5_add_mkey, 1},
    {"use_mkey", kdb5_use_mkey, 1},
    {"list_mkeys", kdb5_list_mkeys, 1},
    {"update_princ_encryption", kdb5_update_princ_encryption, 1},
    {"purge_mkeys", kdb5_purge_mkeys, 1},
    {"tabdump", tabdump, 1},
    {NULL, NULL, 0},
};

static struct _cmd_table *cmd_lookup(name)
    char *name;
{
    struct _cmd_table *cmd = cmd_table;
    while (cmd->name) {
        if (strcmp(cmd->name, name) == 0)
            return cmd;
        else
            cmd++;
    }

    return NULL;
}

#define ARG_VAL (--argc > 0 ? (koptarg = *(++argv)) : (char *)(usage(), NULL))

char **db5util_db_args = NULL;
int    db5util_db_args_size = 0;

static void extended_com_err_fn (const char *myprog, errcode_t code,
                                 const char *fmt, va_list args)
{
    const char *emsg;
    if (code) {
        emsg = krb5_get_error_message (util_context, code);
        fprintf (stderr, "%s: %s ", myprog, emsg);
        krb5_free_error_message (util_context, emsg);
    } else {
        fprintf (stderr, "%s: ", myprog);
    }
    vfprintf (stderr, fmt, args);
    fprintf (stderr, "\n");
}

int add_db_arg(char *arg)
{
    char **temp;
    db5util_db_args_size++;
    temp = realloc(db5util_db_args,
                   sizeof(char *) * (db5util_db_args_size + 1));
    if (temp == NULL)
        return 0;
    db5util_db_args = temp;
    db5util_db_args[db5util_db_args_size-1] = arg;
    db5util_db_args[db5util_db_args_size]   = NULL;
    return 1;
}

int main(argc, argv)
    int argc;
    char *argv[];
{
    struct _cmd_table *cmd = NULL;
    char *koptarg, **cmd_argv;
    char *db_name_tmp = NULL;
    int cmd_argc;
    krb5_error_code retval;

    setlocale(LC_ALL, "");
    set_com_err_hook(extended_com_err_fn);

    /*
     * Ensure that "progname" is set before calling com_err.
     */
    progname = (strrchr(argv[0], '/') ?
                strrchr(argv[0], '/') + 1 : argv[0]);

    retval = kadm5_init_krb5_context(&util_context);
    if (retval) {
        com_err (progname, retval, _("while initializing Kerberos code"));
        exit(1);
    }

    cmd_argv = (char **) malloc(sizeof(char *)*argc);
    if (cmd_argv == NULL) {
        com_err(progname, ENOMEM, _("while creating sub-command arguments"));
        exit(1);
    }
    memset(cmd_argv, 0, sizeof(char *)*argc);
    cmd_argc = 1;

    argv++; argc--;
    while (*argv) {
        if (strcmp(*argv, "-P") == 0 && ARG_VAL) {
            mkey_password = koptarg;
            manual_mkey = TRUE;
        } else if (strcmp(*argv, "-d") == 0 && ARG_VAL) {
            global_params.dbname = koptarg;
            global_params.mask |= KADM5_CONFIG_DBNAME;

            if (asprintf(&db_name_tmp, "dbname=%s", global_params.dbname) < 0)
            {
                com_err(progname, ENOMEM,
                        _("while parsing command arguments"));
                exit(1);
            }

            if (!add_db_arg(db_name_tmp)) {
                com_err(progname, ENOMEM,
                        _("while parsing command arguments\n"));
                exit(1);
            }

        } else if (strcmp(*argv, "-x") == 0 && ARG_VAL) {
            if (!add_db_arg(koptarg)) {
                com_err(progname, ENOMEM,
                        _("while parsing command arguments\n"));
                exit(1);
            }

        } else if (strcmp(*argv, "-r") == 0 && ARG_VAL) {
            global_params.realm = koptarg;
            global_params.mask |= KADM5_CONFIG_REALM;
            /* not sure this is really necessary */
            if ((retval = krb5_set_default_realm(util_context,
                                                 global_params.realm))) {
                com_err(progname, retval,
                        _("while setting default realm name"));
                exit(1);
            }
        } else if (strcmp(*argv, "-k") == 0 && ARG_VAL) {
            if (krb5_string_to_enctype(koptarg, &global_params.enctype)) {
                com_err(progname, EINVAL, _(": %s is an invalid enctype"),
                        koptarg);
                exit(1);
            } else
                global_params.mask |= KADM5_CONFIG_ENCTYPE;
        } else if (strcmp(*argv, "-kv") == 0 && ARG_VAL) {
            global_params.kvno = (krb5_kvno) atoi(koptarg);
            if (global_params.kvno == IGNORE_VNO) {
                com_err(progname, EINVAL, _(": %s is an invalid mkeyVNO"),
                        koptarg);
                exit(1);
            } else
                global_params.mask |= KADM5_CONFIG_KVNO;
        } else if (strcmp(*argv, "-M") == 0 && ARG_VAL) {
            global_params.mkey_name = koptarg;
            global_params.mask |= KADM5_CONFIG_MKEY_NAME;
        } else if (strcmp(*argv, "-sf") == 0 && ARG_VAL) {
            global_params.stash_file = koptarg;
            global_params.mask |= KADM5_CONFIG_STASH_FILE;
        } else if (strcmp(*argv, "-m") == 0) {
            manual_mkey = TRUE;
            global_params.mkey_from_kbd = 1;
            global_params.mask |= KADM5_CONFIG_MKEY_FROM_KBD;
        } else if (cmd_lookup(*argv) != NULL) {
            if (cmd_argv[0] == NULL)
                cmd_argv[0] = *argv;
            else
                usage();
        } else {
            cmd_argv[cmd_argc++] = *argv;
        }
        argv++; argc--;
    }

    if (cmd_argv[0] == NULL)
        usage();

    if( !util_context->default_realm )
    {
        char *temp = NULL;
        retval = krb5_get_default_realm(util_context, &temp);
        if( retval )
        {
            com_err(progname, retval, _("while getting default realm"));
            exit(1);
        }
        krb5_free_default_realm(util_context, temp);
    }

    retval = kadm5_get_config_params(util_context, 1,
                                     &global_params, &global_params);
    if (retval) {
        com_err(progname, retval,
                _("while retreiving configuration parameters"));
        exit(1);
    }

    /*
     * Dump creates files which should not be world-readable.  It is
     * easiest to do a single umask call here.
     */
    (void) umask(077);

    master_keyblock.enctype = global_params.enctype;
    if ((master_keyblock.enctype != ENCTYPE_UNKNOWN) &&
        (!krb5_c_valid_enctype(master_keyblock.enctype))) {
        com_err(progname, KRB5_PROG_KEYTYPE_NOSUPP,
                "while setting up enctype %d", master_keyblock.enctype);
    }

    cmd = cmd_lookup(cmd_argv[0]);
    if (cmd->opendb && open_db_and_mkey())
        return exit_status;

    if (global_params.iprop_enabled == TRUE)
        ulog_set_role(util_context, IPROP_MASTER);
    else
        ulog_set_role(util_context, IPROP_NULL);

    (*cmd->func)(cmd_argc, cmd_argv);

    if( db_name_tmp )
        free( db_name_tmp );

    if( db5util_db_args )
        free(db5util_db_args);

    quit();
    kadm5_free_config_params(util_context, &global_params);
    krb5_free_context(util_context);
    free(cmd_argv);
    return exit_status;
}

/*
 * open_db_and_mkey: Opens the KDC and policy database, and sets the
 * global master_* variables.  Sets dbactive to TRUE if the databases
 * are opened, and valid_master_key to 1 if the global master
 * variables are set properly.  Returns 0 on success, and 1 on
 * failure, but it is not considered a failure if the master key
 * cannot be fetched (the master key stash file may not exist when the
 * program is run).
 */
static int open_db_and_mkey()
{
    krb5_error_code retval;
    krb5_data scratch, pwd, seed;

    dbactive = FALSE;
    valid_master_key = 0;

    if ((retval = krb5_db_open(util_context, db5util_db_args,
                               KRB5_KDB_OPEN_RW | KRB5_KDB_SRV_TYPE_ADMIN))) {
        com_err(progname, retval, _("while initializing database"));
        exit_status++;
        return(1);
    }

    /* assemble & parse the master key name */

    if ((retval = krb5_db_setup_mkey_name(util_context,
                                          global_params.mkey_name,
                                          global_params.realm,
                                          &mkey_fullname, &master_princ))) {
        com_err(progname, retval, _("while setting up master key name"));
        exit_status++;
        return(1);
    }
    if ((retval = krb5_db_get_principal(util_context, master_princ, 0,
                                        &master_entry))) {
        com_err(progname, retval, _("while retrieving master entry"));
        exit_status++;
        (void) krb5_db_fini(util_context);
        return(1);
    }

    if (global_params.mask & KADM5_CONFIG_KVNO)
        master_kvno = global_params.kvno; /* user specified */
    else
        master_kvno = IGNORE_VNO;

    /* the databases are now open, and the master principal exists */
    dbactive = TRUE;

    if (mkey_password) {
        pwd.data = mkey_password;
        pwd.length = strlen(mkey_password);
        retval = krb5_principal2salt(util_context, master_princ, &scratch);
        if (retval) {
            com_err(progname, retval, _("while calculated master key salt"));
            exit_status++;
            return(1);
        }

        /* If no encryption type is set, use the default */
        if (master_keyblock.enctype == ENCTYPE_UNKNOWN)
            master_keyblock.enctype = DEFAULT_KDC_ENCTYPE;
        if (!krb5_c_valid_enctype(master_keyblock.enctype))
            com_err(progname, KRB5_PROG_KEYTYPE_NOSUPP,
                    "while setting up enctype %d",
                    master_keyblock.enctype);

        retval = krb5_c_string_to_key(util_context, master_keyblock.enctype,
                                      &pwd, &scratch, &master_keyblock);
        if (retval) {
            com_err(progname, retval,
                    _("while transforming master key from password"));
            exit_status++;
            return(1);
        }
        free(scratch.data);
        mkey_password = 0;

    } else {
        if ((retval = krb5_db_fetch_mkey(util_context, master_princ,
                                         master_keyblock.enctype,
                                         manual_mkey, FALSE,
                                         global_params.stash_file,
                                         &master_kvno,
                                         0, &master_keyblock))) {
            com_err(progname, retval, _("while reading master key"));
            com_err(progname, 0, _("Warning: proceeding without master key"));
            exit_status++;
            return(0);
        }
    }

    if ((retval = krb5_db_fetch_mkey_list(util_context, master_princ,
                                          &master_keyblock))) {
        com_err(progname, retval, "while getting master key list");
        com_err(progname, 0, "Warning: proceeding without master key list");
        exit_status++;
        return(0);
    }

    seed.length = master_keyblock.length;
    seed.data = (char *) master_keyblock.contents;

    if ((retval = krb5_c_random_seed(util_context, &seed))) {
        com_err(progname, retval, _("while seeding random number generator"));
        exit_status++;
        memset(master_keyblock.contents, 0, master_keyblock.length);
        krb5_free_keyblock_contents(util_context, &master_keyblock);
        return(1);
    }

    if (global_params.iprop_enabled) {
        if (ulog_map(util_context, global_params.iprop_logfile,
                     global_params.iprop_ulogsize)) {
            fprintf(stderr, _("%s: Could not map log\n"), progname);
            exit_status++;
            return(1);
        }
    }

    valid_master_key = 1;
    dbactive = TRUE;
    return 0;
}

#ifdef HAVE_GETCWD
#undef getwd
#endif

int
quit()
{
    krb5_error_code retval;
    static krb5_boolean finished = 0;

    if (finished)
        return 0;
    ulog_fini(util_context);
    retval = krb5_db_fini(util_context);
    zapfree(master_keyblock.contents, master_keyblock.length);
    krb5_free_principal(util_context, master_princ);
    finished = TRUE;
    if (retval && retval != KRB5_KDB_DBNOTINITED) {
        com_err(progname, retval, _("while closing database"));
        exit_status++;
        return 1;
    }
    return 0;
}

static void
add_random_key(argc, argv)
    int argc;
    char **argv;
{
    krb5_error_code ret;
    krb5_principal princ;
    krb5_db_entry *dbent;
    krb5_timestamp now;

    krb5_key_salt_tuple *keysalts = NULL;
    krb5_int32 num_keysalts = 0;

    int free_keysalts;
    char *me = progname;
    char *ks_str = NULL;
    char *pr_str;
    krb5_keyblock *tmp_mkey;

    if (argc < 2)
        usage();
    for (argv++, argc--; *argv; argv++, argc--) {
        if (!strcmp(*argv, "-e")) {
            argv++; argc--;
            ks_str = *argv;
            continue;
        } else
            break;
    }
    if (argc < 1)
        usage();
    pr_str = *argv;
    ret = krb5_parse_name(util_context, pr_str, &princ);
    if (ret) {
        com_err(me, ret, _("while parsing principal name %s"), pr_str);
        exit_status++;
        return;
    }
    ret = krb5_db_get_principal(util_context, princ, 0, &dbent);
    if (ret) {
        com_err(me, ret, _("while fetching principal %s"), pr_str);
        exit_status++;
        return;
    }
    ret = krb5_string_to_keysalts(ks_str,
                                  NULL, NULL, 0,
                                  &keysalts,
                                  &num_keysalts);
    if (ret) {
        com_err(me, ret, _("while parsing keysalts %s"), ks_str);
        exit_status++;
        return;
    }
    if (!num_keysalts || keysalts == NULL) {
        num_keysalts = global_params.num_keysalts;
        keysalts = global_params.keysalts;
        free_keysalts = 0;
    } else
        free_keysalts = 1;

    /* Find the mkey used to protect the existing keys */
    ret = krb5_dbe_find_mkey(util_context, dbent, &tmp_mkey);
    if (ret) {
        com_err(me, ret, _("while finding mkey"));
        krb5_db_free_principal(util_context, dbent);
        exit_status++;
        return;
    }

    ret = krb5_dbe_ark(util_context, tmp_mkey, keysalts, num_keysalts, dbent);
    if (free_keysalts)
        free(keysalts);
    if (ret) {
        com_err(me, ret, "while randomizing principal %s", pr_str);
        krb5_db_free_principal(util_context, dbent);
        exit_status++;
        return;
    }
    dbent->attributes &= ~KRB5_KDB_REQUIRES_PWCHANGE;
    ret = krb5_timeofday(util_context, &now);
    if (ret) {
        com_err(me, ret, _("while getting time"));
        krb5_db_free_principal(util_context, dbent);
        exit_status++;
        return;
    }
    ret = krb5_dbe_update_last_pwd_change(util_context, dbent, now);
    if (ret) {
        com_err(me, ret, _("while setting changetime"));
        krb5_db_free_principal(util_context, dbent);
        exit_status++;
        return;
    }
    ret = krb5_db_put_principal(util_context, dbent);
    krb5_db_free_principal(util_context, dbent);
    if (ret) {
        com_err(me, ret, _("while saving principal %s"), pr_str);
        exit_status++;
        return;
    }
    printf(_("%s changed\n"), pr_str);
}
